### PR TITLE
makes controller methods virtual

### PR DESCRIPTION
**Pull request by efernandez**
To my best knowledge, this methods should be virtual.
See the imu_sensor_controller for instance.

### CODE REVIEW

**Comment by po1** [comment]
That is correct. They _should_, but don't _have to_.
Only the parent class methods must be declared `virtual`. The child class methods should be declared `virtual` as an indication to the reader.

The `virtual` qualifier is targeting sub-classes. If nobody inherits from `DiffDriveController`, then it is just for decoration.

That being said, just because other controllers do it so, +1

File: diff_drive_controller/include/diff_drive_controller/diff_drive_controller.h [hunk 1]
```diff
@@ -72,28 +72,28 @@
      * \param root_nh       Node handle at root namespace
      * \param controller_nh Node handle inside the controller namespace
      */
-    bool init(hardware_interface::VelocityJointInterface* hw,
-              ros::NodeHandle& root_nh,
-              ros::NodeHandle &controller_nh);
+    virtual bool init(hardware_interface::VelocityJointInterface* hw,
+        ros::NodeHandle& root_nh,
```
Comment (bmagyar) [inline diff_drive_controller/include/diff_drive_controller/diff_drive_controller.h]: can you fix the indent here and on the next one?

**Comment by efernandez** [comment]
@bmagyar I know you'll like the indentation! ;)

**Comment by bmagyar** [comment]
Hehe, yes. (btw sorry for closing and opening the issue, my touchpad went crazy)

+1

**Comment by adolfo-rt** [comment]
Late to the conversation.

The `virtual` qualifier should in fact not be present in leaf classes, because there will be no childs reimplementing it. In the  `imu_sensor_controller`, I copy-pasted the structure from the existing `joint_states_controller`, so my bad.

**Comment by po1** [comment]
@adolfo-rt if this means that there is an intention of making it explicit that these classes should be final and not derived from, then I withdraw my +1.

:-1:  :smiling_imp:

**Comment by adolfo-rt** [comment]
My statement still has a [source needed]. I just haven't found the nice read that explained it.

**Comment by efernandez** [comment]
This PR shouldn't be applied.
It seems that I can't reject it; could somebody do it?

**Pull request closed without merging**